GCP Deploy guide update (#2631)
withastro/docs: src/pages/en/guides/deploy/google-cloud.mdx

```docker title="Dockerfile"
FROM node:lts as runtime
WORKDIR /app

COPY . .

RUN npm install
RUN npm run build

ENV HOST=0.0.0.0
ENV PORT=3000
EXPOSE 3000
CMD node ./dist/server/entry.mjs
```

:::tip[Keep this in mind]
These are just examples of Dockerfiles. You can customize them to your needs. For example, you could use another image, like `node:lts-alpine`.

```docker title=".Dockerfile del={1} add={2}
FROM node:lts as runtime
FROM node:lts-alpine as runtime
```
:::

#### Build the Container

Once the Dockerfile is created, build it into an image and push it to Google Cloud. There are a few ways to accomplish this:

**Build locally using Docker**:

Use the `docker build` command to build the image, `docker tag` to give it a tag, then `docker push` to push it to a registry. In the case of Google Cloud, [`Google Cloud Registry`](https://cloud.google.com/container-registry/docs/pushing-and-pulling#add-registry) is the easiest option, but you can also use [Docker Hub](<https://hub.docker.com/>).

```bash
# build your container
docker build .

docker tag SOURCE_IMAGE HOSTNAME/PROJECT-ID/TARGET-IMAGE:TAG

# Push your image to a registry
docker push HOSTNAME/PROJECT-ID/IMAGE:TAG
```

> :::note[Replace variables]
> Replace the following:
> 
> `SOURCE_IMAGE` is the local image name or image ID.
> `HOSTNAME` is the registry host (`gcr.io`, `eu.grc.io`, `asia.gcr.io`, `us.gcr.io`).
> `PROJECT` is your Google Cloud project ID.
> `TARGET-IMAGE` is the name for the image when it's stored in the registry.
> `TAG` is the version associated with the image
> :::
> 

Reviewer: ````suggestion
Use the following values from your own project in your commands:

- `SOURCE_IMAGE`: the local image name or image ID.

- `HOSTNAME`: the registry host (`gcr.io`, `eu.grc.io`, `asia.gcr.io`, `us.gcr.io`).

- `PROJECT`: your Google Cloud project ID.

- `TARGET-IMAGE`: the name for the image when it's stored in the registry.

- `TAG` is the version associated with the image

````
Removing the note and using bullet points for readability! Not entirely sold on the intro phrase, but I feel like it should refer to the commands somehow.

Author: What do you think of this intro phrase? 

**"Change the following values in the commands above to match your project:"**

Reviewer 2: I'm happy with that!

Deploying a project to Google Cloud requires some UI clicks. (Most of these actions can also be done using the [gcloud CLI](https://cloud.google.com/sdk/gcloud/)).

### Cloud Run (SSR and static)

Cloud Run is a serverless platform that allows you to run a container without having to manage any infrastructure. It can be used to deploy both static and SSR sites.

#### 1. Preparing the service

1. Create a new GCP project, or select one you already have.

2. Make sure the Cloud Run API is enabled.

3. Create a new service.

> 4. Make an empty `Dockerfile` in the root of your project.

Reviewer 2: ```suggestion
4. Make an empty `Dockerfile` in the root of your project.
```
Can you double-check that it isn't supposed to be `.Dockerfile` with the dot?

Author: Yes, it's `Dockerfile`

source: <https://docs.docker.com/engine/reference/builder/>

#### 2. Building the container

There are a couple of ways to build the container:

- Locally using Docker
- Using [Cloud Build](https://cloud.google.com/build)

> - Using any other CI/CD tool (e.g. [GitHub Actions](<https://docs.github.com/en/actions/publishing-packages/publishing-docker-images>))

Author: No, at this moment, I've not included any of them actually. I think configuring and setting up pipelines, is a little bit outside of the scope of the deployment guide, but if you'd like, I can still add examples or better resources.

In this PR I focussed more on just writing and building a container. Deploying to a cloud platform can be done in so many different ways which all could be correct depending on the usecase.

@Jutanium maybe it's a better Idea to split the deployment guide into a more generic "building with Docker" and "Deploying to Google Cloud"

All three of the options with SSR, require you to add a `Dockerfile`

## Local

Requires you to build with `docker build` command and push to a remote using `docker push`. (in this case that would be Artifact registry or container registry on Google Cloud)

## Cloudbuild

To setup a pipeline with CloudBuild, you have to add a separate `cloudbuild.yaml` file with a description what the pipeline should do. (build and push container, then create new Cloud Run service)

> - Using any other CI/CD tool (e.g. [GitHub Actions](<https://docs.github.com/en/actions/publishing-packages/publishing-docker-images>))

Reviewer 2: Gotcha - I see how the details can be out of the scope of this page, so I'll defer to your judgement. We can:
- Leave as is
- Add those **local** and **cloudbuild** details you just laid out as sub-bullet points
- Hyperlink the bullets to external guides

> - Using any other CI/CD tool (e.g. [GitHub Actions](<https://docs.github.com/en/actions/publishing-packages/publishing-docker-images>))

Author: I updated it with extra bullet points. I think this makes it more clear! Thanks for the suggestion!

##### Dockerfile for static sites (example)

This following Dockerfile will build your site and serve it using Apache HTTPD (default config) on port `80`.

```docker title="Dockerfile"
FROM node:lts as build
WORKDIR /app
COPY . .
RUN npm i
RUN npm run build

FROM httpd:2.4 as runtime
COPY --from=build /app/dist /usr/local/apache2/htdocs/
EXPOSE 80
```

:::note[notes]

> Keep in mind you might need to change `node:lts` to a pinned version, if you're using another package manager than `npm`, you might need to install it and change the scripts.

Author: **"In what situations would we want to change node:lts to a pinned version?"**:

Well, a simple example could be; Imagine the Node LTS versions changing from 16 to 18 (not that long ago). If you just have `node:lts` set, you might experience breakage if Astro (or any other package) does not support that version yet

**"If we're using another package manager than npm, we just need to replace npm i and npm run build in this file, right?"**

Yes and no, depending on the image. For example, the standard `Node:lts` does include `yarn` & `npm`, but not `pnpm`. So if you want to use `npm`, you could do something like:

```docker
FROM node:lts as build
RUN npm i -g pnpm@latest
RUN pnpm i
RUN pnpm build
```

> Keep in mind you might need to change `node:lts` to a pinned version, if you're using another package manager than `npm`, you might need to install it and change the scripts.

Reviewer 2: I like including that example so it's clear what you mean.
Also, I believe we can leave out the `pinned version` point here because we make it in the following section.

````suggestion
If you're using a package manager other than `npm`, you'll need to change the commands:

```docker title="Dockerfile" add={4} "pnpm"
FROM node:lts as build
WORKDIR /app
COPY . .
RUN npm i -g pnpm@latest
RUN pnpm i
RUN pnpm run build
...
```

````

:::

##### Dockerfile for SSR sites (example)

This following Dockerfile will build your site and serve it using Node.js (SSR adapter) on port `3000`.

```docker title="Dockerfile"
FROM node:lts as runtime
WORKDIR /app

COPY . .

RUN npm install
RUN npm run build

ENV HOST=0.0.0.0
ENV PORT=3000
EXPOSE 3000
CMD node ./dist/server/entry.mjs
```

:::tip[Keep this in mind]

> These are just examples of Dockerfiles. You can customize them to your needs. For example, using `node:lts-alpine`

Reviewer 2: ````suggestion
These are just examples of Dockerfiles. You can customize them to your needs. For example, you could use another image, like `node:lts-alpine`.

```docker title=".Dockerfile del={1} add={2}
FROM node:lts as runtime
FROM node:lts-alpine as runtime
```
````

:::

#### 3. Deploying the container

> Deployment can be handled manually in terminal or automatically using [Cloud Build](https://cloud.google.com/build) or any other CI/CD system.

Reviewer 2: ```suggestion
Deployment can be handled manually in your terminal or automatically using [Cloud Build](https://cloud.google.com/build) or any other CI/CD system.
```
Should we add an example of deploying from your terminal?

Author: Hmm, I don't know if it's a good idea, kinda the same as the first comment. Of course, I'm willing to add it if you guys think it is better or more clear.

Reviewer 2: I understand now why it's out-of-scope!

Deploying a project to Google Cloud requires some UI clicks. (Most of these actions can also be done using the [gcloud CLI](https://cloud.google.com/sdk/gcloud/)).

### Cloud Run (SSR and static)

Cloud Run is a serverless platform that allows you to run a container without having to manage any infrastructure. It can be used to deploy both static and SSR sites.

#### 1. Preparing the service

1. Create a new GCP project, or select one you already have.

2. Make sure the Cloud Run API is enabled.

3. Create a new service.

4. Make an empty `Dockerfile` in the root of your project.

#### 2. Building the container

There are a couple of ways to build the container:

> - Locally using Docker

Author: @sarah11918, I rewrote the way it's 1. structured and 2. explain it (I hope) a bit more clear

  - Requires you to build with `docker build` command and push to a remote using `docker push`. (That would be Artifact registry or container registry on Google Cloud)
- Using [Cloud Build](https://cloud.google.com/build)
  - To setup a pipeline with CloudBuild, you have to add a separate `cloudbuild.yaml` file with a description what the pipeline should do. [Check the Google Cloud Build documentation here](<https://cloud.google.com/build/docs/build-push-docker-image>).
- Using any other CI/CD tool (e.g. [GitHub Actions](<https://docs.github.com/en/actions/publishing-packages/publishing-docker-images>))

> ##### Dockerfile for static sites (example)

Author: Well, it's actually a little more complicated. All of these options are Docker. But inside of Docker, you can choose to run a static site (first h5) or with a NodeJS Server (second h5). The 3 options you are referring to only address the different ways you can build (the same) Docker container

> ##### Dockerfile for static sites (example)

Reviewer: OK, then I think this needs to be "explain it like I'm Sarah" clearer. 😆 

There are 3 options above to build a DOCKER container, and all of them require this file?

> ##### Dockerfile for static sites (example)

Author: Yes! Exactly! Docker can be used in a lot of different ways. This step only addresses building a container. That container, you can define how you want, but always requires a `Dockerfile`.

That's why I listed two examples of possible `Dockerfile`'s
